pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/OpChainSchedulerService.java

```diff
@@ -31,151 +33,84 @@
-            // throw if the operatorChain is cancelled.
-            if (_cancelledRequests.asMap().containsKey(operatorChain.getId().getRequestId())) {
-              throw new InterruptedException("Query was cancelled!");
-            }
             operatorChain.getStats().executing();
-            // so long as there's work to be done, keep getting the next block
-            // when the operator chain returns a NOOP block, then yield the execution
-            // of this to another worker
             TransferableBlock result = operatorChain.getRoot().nextBlock();
-            while (!result.isNoOpBlock() && !result.isEndOfStreamBlock()) {
+            while (!result.isEndOfStreamBlock()) {
               result = operatorChain.getRoot().nextBlock();
             }
-
-            if (result.isNoOpBlock()) {
-              // TODO: There should be a waiting-for-data state in OpChainStats.
-              operatorChain.getStats().queued();
-              _scheduler.yield(operatorChain);
+            isFinished = true;
+            if (result.isErrorBlock()) {
+              returnedErrorBlock = result;
+              LOGGER.error("({}): Completed erroneously {} {}", operatorChain, operatorChain.getStats(),
+                  result.getDataBlock().getExceptions());
             } else {
-              isFinished = true;
-              if (result.isErrorBlock()) {
-                returnedErrorBlock = true;
-                LOGGER.error("({}): Completed erroneously {} {}", operatorChain, operatorChain.getStats(),
-                    result.getDataBlock().getExceptions());
-              } else {
-                LOGGER.debug("({}): Completed {}", operatorChain, operatorChain.getStats());
-              }
+              LOGGER.debug("({}): Completed {}", operatorChain, operatorChain.getStats());
             }
           } catch (Exception e) {
             LOGGER.error("({}): Failed to execute operator chain! {}", operatorChain, operatorChain.getStats(), e);
             thrown = e;
           } finally {
-            if (returnedErrorBlock || thrown != null) {
+            if (returnedErrorBlock != null || thrown != null) {
+              if (thrown == null) {
+                String blockMsg;
+                if (returnedErrorBlock.getDataBlock() instanceof MetadataBlock) {
+                  MetadataBlock metadataBlock = (MetadataBlock) returnedErrorBlock.getDataBlock();
+                  blockMsg = String.join(", ", metadataBlock.getExceptions().values());
+                } else {
+                  blockMsg = "Unknown";
+                }
+                thrown = new RuntimeException("Error block " + blockMsg);
+              }
               cancelOpChain(operatorChain, thrown);
             } else if (isFinished) {
               closeOpChain(operatorChain);
             }
           }
         }
       });
-    }
+    _submittedOpChainMap.put(operatorChain.getId(), scheduledFuture);
```

Comment:
We need to remove the entry when closing/cancelling the `OpChain`

Reply:
Done in 2e4bde